Add Doctor command to create missing linked notes

hikchoi opened the pull request:
This PR:
- adds a new Doctor command that goes through wikilinks in all notes and create notes that are linked but does not exist yet.

related to: #355

kevinslin commented on packages/plugin-core/src/commands/Doctor.ts:
```diff
@@ -67,6 +96,71 @@ export class DoctorCommand extends BasicCommand<CommandOpts, CommandOutput> {
+          const wsRoot = DendronWorkspace.wsRoot();
+          wildWikiLinks = wildWikiLinks.concat(
+            _.filter(links, (link) => {
+              if (link.type != "wiki") {
```
since its javascript, we want to use strict equality check `!==` (see [here](https://stackoverflow.com/questions/359494/which-equals-operator-vs-should-be-used-in-javascript-comparisons))

kevinslin commented on packages/plugin-core/src/commands/Doctor.ts:
```diff
@@ -67,6 +96,71 @@ export class DoctorCommand extends BasicCommand<CommandOpts, CommandOutput> {
+          // show preview before creating
+          await this.showPreview(uniqueCandidates);
+          const options = ["proceed", "cancel"];
+          const shouldProceed = await VSCodeUtils.showQuickPick(options, {
```
check for `undefined` here, it means the user pressed escape in which case we should cancel the command

hikchoi replied:
wouldn't `shouldProceed !== "proceed"` check for `undefined` as well? If I run press escape it will cancel the command and show the information message correctly.

kevinslin replied:
your right, i missed that!

kevinslin commented on packages/plugin-core/src/test/suite-integ/DoctorCommand.test.ts:
```diff
@@ -266,3 +270,37 @@ suite("notes", function () {
+          })
+        );
+        await cmd.run();
+        done();
```
don't forget to check if the note is created

hikchoi replied:
ah yeah. this test file isn't complete 😄  I pushed it with the code mid way

kevinslin commented on packages/plugin-core/src/test/suite-integ/DoctorCommand.test.ts:
```diff
@@ -266,3 +270,37 @@ suite("notes", function () {
+    },
+  });
+
+  test.only("basic", (done) => {
```
don't forget to remove `.only` in the final pull request

kevinslin commented on packages/plugin-core/src/test/suite-integ/DoctorCommand.test.ts:
```diff
@@ -11,7 +15,7 @@ import { DoctorCommand } from "../../commands/Doctor";
-import { setupBeforeAfter } from "../testUtilsV3";
+import { runLegacySingleWorkspaceTest, setupBeforeAfter } from "../testUtilsV3";
 
 suite("notes", function () {
```
this was my mistake but if you could change the name here to `DoctorComamndTest` :)

kevinslin commented on packages/plugin-core/src/commands/Doctor.ts:
```diff
@@ -67,6 +96,71 @@ export class DoctorCommand extends BasicCommand<CommandOpts, CommandOutput> {
         });
         break;
       }
+      case DoctorActions.CREATE_MISSING_LINKED_NOTES: {
```
can we have the bulk of this logic in `DoctorCLICommand`? that way, it will be accessible in the CLI as well. for the CLI, we might add a `dry-run` flag in place of the preview

hikchoi replied:
Do you think it's better to make this a separate command (like `BackfillV2Command` for `DoctorActions.FIX_FRONTMATTER` does)?

or should this be a cli-first command?

flammehawk replied:
Just my two cents, but wouldn't it be better if cli, plugin, or API endpoint on the engine are just different front ends for the Command? 
Since it should be executable from the cli,  plugin, or an API call to the engine? 

The plugin could call the API to get it done, cli could execute it without needing the engine or call the engine and format the result differently for a console.

kevinslin replied:
the CLI, plugin and API endpoint are just different frontends for the command. 
the dendron engine handles the core operations (CRUD and a few extras) and it is accessed through the API.

just published an updated startup diagram [here](https://wiki.dendron.so/notes/c160ddce-edec-4f6e-841b-418d6030fa37.html)

hikchoi commented:
- moved the core logic to cli.
- handled case where multiple vaults had wild links.
- handled cross-vault wild link cases.
- added cli and plugin side tests for:
  - basic functionality
  - on cancel / dry run
  - multiple vault case
  - cross-vault case
  - aliased wild links
 
I was planning to add tsdoc comments for the new feature but I am deferring this to a separate PR to favor a timely delivery.

kevinslin commented on packages/dendron-cli/src/commands/doctor.ts:
```diff
@@ -66,6 +72,43 @@ export class DoctorCLICommand extends CLICommand<CommandOpts, CommandOutput> {
+            vault: destVault,
+            notes: notes,
+            wsRoot: wsRoot,
+          }) as NoteProps;
```
don't cast it here because it could be undefined. the !noteExists should still be valid without a cast

kevinslin commented on packages/dendron-cli/src/commands/doctor.ts:
```diff
@@ -66,6 +72,43 @@ export class DoctorCLICommand extends CLICommand<CommandOpts, CommandOutput> {
   }
 
+  getWildLinkDestinations(notes: NoteProps[], wsRoot: string) {
+    let wildWikiLinks = [] as DLink[];
```
instead of casting, specify in assignment

```
let wildWikiLinks: DLink[] = [];
```

kevinslin commented on packages/dendron-cli/src/commands/doctor.ts:
```diff
@@ -66,6 +72,43 @@ export class DoctorCLICommand extends CLICommand<CommandOpts, CommandOutput> {
+          vault: destVault,
+        };
+      }
+    ) as NoteProps[];
```
move type to var decalration

```
uniqueCandidates: NoteProps[] = _.map
```

kevinslin reviewed (changes requested):
nice work. just some typescript type fixes but otherwise, this looks great!

hikchoi commented:
pushed the suggested changes. Still a bit of resistance coming from mainly using a non-typed language for a few years 😅
